### PR TITLE
Fix session tracker count raising a warning on PHP 7.2

### LINKED ISSUE #557: PHP 7.2: count(): Parameter must be an array or an object that implements Countable
### Description
*count(): Parameter must be an array or an object that implements Countable* warning is raised.

### Issue
```php
        $sessions = $this->getSessionCounts();
        $this->setSessionCounts([]);
        if (count($sessions) == 0) {
            return;
        }
```
in bugsnag/bugsnag/src/SessionTracker.php:285
### Environment

* bugsnag-php version: 3.19.0
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by imjoehaines**
## Goal

Fix a warning being raised when a session storage function returns `null` on PHP versions >=7.2 — see #557

### Added

- More tests for `SessionTracker`

### Changed

- `SessionTracker::deliverSessions` now checks if the return value from the session storage function is an array before calling `count` on it
- `SessionTracker::incrementSessions` now checks the return value from the session storage function is an array before calling `array_key_exists` on it
- `SessionTracker::incrementSessions` also checks the return value for "last sent" is an integer before doing maths with it, though we specifically cast `null` to `0` to keep the previous behaviour with `null` values
    I think this is a safe change because our code should be the only place this value is set (`setLastSent`), so this should only be a change in behaviour for incorrectly implemented `storageFunction` callbacks
- `SessionTracker` didn't import `Exception` or `InvalidArgumentException`
    This means none of the `setXFunction` methods would throw if given invalid parameters (they'd crash due to no `Bugsnag\InvalidArgumentException` class existing instead)
    It also means we wouldn't catch exceptions when `deliverSessions` was called, so the retry functionality wasn't working correctly
- Some of the namespaces for returned Guzzle `Client` objects have been updated
- StyleCI is failing on master so the suggested fixes have been applied on this branch (in 1364246ff0316ebe4114c93b1059d0bdd20e34de)

## Tests

- I've added a test for non-array return values for `getSessionCounts` to make sure the fix is working — these will fail if the warning was still being raised
- I improved the tests to avoid mocking the `SessionTracker` class itself, as it's the class being tested
- I've also added a test for when `startSession` is called, which is mostly the same as `sendSessions` but the session is pre-populated

### Outstanding Questions

- The return value of `Configuration::getSessionClient` is `GuzzleHttp\ClientInterface` which no longer has a `post` method available (as of Guzzle v6)
    This means we need to check whether it exists or not, to know which version we're on and therefore which method to call
    We could change the return type to the concrete `GuzzleHttp\Client` class instead, which has a `post` method in v7, v6 and v5 of Guzzle

### Linked issues

Fixes #557 
Supersedes https://github.com/bugsnag/bugsnag-laravel/pull/378 and https://github.com/bugsnag/bugsnag-php/pull/561

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: src/SessionTracker.php [hunk 1]
```diff
@@ -281,27 +286,36 @@ protected function constructPayload($sessions)
     protected function deliverSessions()
     {
         $sessions = $this->getSessionCounts();
+
         $this->setSessionCounts([]);
-        if (count($sessions) == 0) {
+
+        if (!is_array($sessions) || count($sessions) === 0) {
```
Comment (imjoehaines) [inline src/SessionTracker.php:292]: This `!is_array` check fixes the issue. It doesn't seem worth adding `Countable` support here, but we can do if that's desirable

File: src/SessionTracker.php [hunk 2]
```diff
@@ -281,27 +286,36 @@ protected function constructPayload($sessions)
...
         try {
-            $http->post('', [
-                'json' => $payload,
-                'headers' => $headers,
-            ]);
+            if (method_exists($http, 'request')) {
+                $http->request('POST', '', $options);
+            } else {
+                $http->post('', $options);
+            }
```
Comment (imjoehaines) [inline src/SessionTracker.php:318]: This isn't ideal but the return type of `Configuration::getSessionClient` is `GuzzleHttp\ClientInterface`, which doesn't have a `post` method in version 6 (it's defined on the concrete class instead)

We can tidy this up by depending on the concrete `GuzzleHttp\Client` class instead, but we have similar checks to this elsewhere so I've not done that as part of this PR

File: src/SessionTracker.php (hunk 1 above)
Comment (imjoehaines) [reply src/SessionTracker.php:292]: Actually we can't support `Countable` — we call `array_key_exists` elsewhere so we explicitly only support arrays here

Comment (GrahamCampbell) [reply src/SessionTracker.php:296]: It would help to add phpdoc with type information to all the methods in this class, so the return type of `getSessionCounts` is clear. That function should probably throw away anything that is not null or an array, already.

File: src/SessionTracker.php [hunk 3]
```diff
@@ -281,27 +290,36 @@ protected function constructPayload($sessions)
...
         ];
+
         $this->setLastSent();
 
         try {
-            $http->post('', [
-                'json' => $payload,
-                'headers' => $headers,
-            ]);
+            if (method_exists($http, 'request')) {
```
Comment (GrahamCampbell) [inline src/SessionTracker.php:318]: This code should match https://github.com/bugsnag/bugsnag-php/commit/3264e0da5c4082337659870752cffa16a4e5920e#diff-add00846710b59ac0af2a07e1ede2747R209-R225. `method_exists` is technically wrong because it could detect something which is not a public method.

Comment (imjoehaines) [reply src/SessionTracker.php:318]: Just making sure I understand — you mean check the `ClientInterface` specifically, rather than the `$http` object?

File: src/SessionTracker.php (hunk 1 above)
Comment (imjoehaines) [reply src/SessionTracker.php]: I've made this change in https://github.com/bugsnag/bugsnag-php/pull/579/commits/d2daba09ecd4d881a16e10291ae5304e57751153

I think it's safe because it's essentially the same as what we were doing internally anyway, but `getCurrentSession` is public so it could be a BC break. I'm not convinced it is though, given the bug we're fixing here is caused by returning the wrong type from the custom functions, so if anyone was relying on that it's likely they were triggering this bug (or a similar one with `lastSent`) anyway

File: src/SessionTracker.php (hunk 3 above)
Comment (GrahamCampbell) [reply src/SessionTracker.php]: Yes. This is because the interface tells you about the public API, rather than not necessarily public class methods.
